File: Algorithms/src/120. Triangle/Code/solution.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int n = triangle.size();
        // dp array initialized to last row values
        int[] dp = new int[n];
        List<Integer> last = triangle.get(n - 1);
        for (int i = 0; i < n; i++) dp[i] = last.get(i);

        // iterate from second-last row up
        for (int i = n - 2; i >= 0; --i) {
            for (int j = 0; j <= i; ++j) {
                dp[j] = triangle.get(i).get(j) + Math.min(dp[j], dp[j + 1]);
            }
        }
        return dp[0];
    }
}

File: Algorithms/src/120. Triangle/Code/solution.py
from typing import List

class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        n = len(triangle)
        # dp initialized with last row (copy)
        dp = triangle[-1].copy()

        # iterate from second-last row up to the top
        for i in range(n - 2, -1, -1):
            for j in range(i + 1):
                dp[j] = triangle[i][j] + min(dp[j], dp[j + 1])

        return dp[0]
